pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java

```diff
@@ -44,134 +43,126 @@
-    if (tableConfig != null) {
-      TableTaskConfig taskConfig = tableConfig.getTaskConfig();
-      if (taskConfig != null) {
-        for (String taskType : _taskTypes) {
-          if (taskConfig.isTaskTypeEnabled(taskType)) {
-            _enabledTableConfigMap.get(taskType).add(tableConfig);
+    // Scan all table configs to get the tables with tasks enabled
+    for (String tableNameWithType : tableNamesWithType) {
+      TableConfig tableConfig = _pinotHelixResourceManager.getTableConfig(tableNameWithType);
+      if (tableConfig != null) {
+        TableTaskConfig taskConfig = tableConfig.getTaskConfig();
+        if (taskConfig != null) {
+          for (String taskType : taskTypes) {
+            if (taskConfig.isTaskTypeEnabled(taskType)) {
+              enabledTableConfigMap.get(taskType).add(tableConfig);
+            }
           }
         }
       }
     }
-  }
-
-  @Override
-  protected void exceptionHandler(String tableNameWithType, Exception e) {
-    LOGGER.error("Exception in PinotTaskManager for table {}", tableNameWithType, e);
-  }
 
-  @Override
-  protected void postprocess() {
     // Generate each type of tasks
-    _tasksScheduled = new HashMap<>(_numTaskTypes);
-    for (String taskType : _taskTypes) {
+    Map<String, String> tasksScheduled = new HashMap<>(numTaskTypes);
+    for (String taskType : taskTypes) {
       LOGGER.info("Generating tasks for task type: {}", taskType);
       PinotTaskGenerator pinotTaskGenerator = _taskGeneratorRegistry.getTaskGenerator(taskType);
-      List<PinotTaskConfig> pinotTaskConfigs = pinotTaskGenerator.generateTasks(_enabledTableConfigMap.get(taskType));
+      List<PinotTaskConfig> pinotTaskConfigs = pinotTaskGenerator.generateTasks(enabledTableConfigMap.get(taskType));
       int numTasks = pinotTaskConfigs.size();
       if (numTasks > 0) {
         LOGGER
             .info("Submitting {} tasks for task type: {} with task configs: {}", numTasks, taskType, pinotTaskConfigs);
-        _tasksScheduled.put(taskType, _helixTaskResourceManager
+        tasksScheduled.put(taskType, _helixTaskResourceManager
             .submitTask(pinotTaskConfigs, pinotTaskGenerator.getNumConcurrentTasksPerInstance()));
-        _metricsRegistry.addMeteredTableValue(taskType, ControllerMeter.NUMBER_TASKS_SUBMITTED, numTasks);
+        _controllerMetrics.addMeteredTableValue(taskType, ControllerMeter.NUMBER_TASKS_SUBMITTED, numTasks);
       }
     }
-  }
 
-  /**
-   * Returns the tasks that have been scheduled as part of the postprocess
-   * @return
-   */
-  private Map<String, String> getTasksScheduled() {
-    return _tasksScheduled;
+    return tasksScheduled;
   }
 
-  /**
-   * Performs necessary cleanups (e.g. remove metrics) when the controller leadership changes.
-   */
-  @Override
-  public void stopTask() {
-    LOGGER.info("Perform task cleanups.");
-    // Performs necessary cleanups for each task type.
+  private void nonLeaderCleanUp() {
```

Comment:
Can some comments be added, either here or above cleanupTask explaining why non leader cleanup is needed in this task? Since these tasks run on controller leadership gain, it's confusing to see this method here

Reply:
Added comments in BasePeriodicTask